Sources/Basic/ArgumentParser.swift

```diff
@@ -0,0 +1,264 @@
+public protocol ArgumentKind {
+    /// This will be called when the option is encountered while parsing.
+    /// To get the argument after the option call popArg closure which might
+    /// fail if there is no next entry.
+    init(popArg: () throws -> (String)) throws
+
+    /// This will be called for positional arguments with the value discovered.
+    init?(arg: String)
+}
+
+// MARK:- ArgumentKind conformance for common types
+
+extension String: ArgumentKind {
+    public init?(arg: String) {
+        self = arg
+    }
+
+    public init(popArg: () throws -> (String)) throws {
+        self = try popArg()
+    }
+}
+
+extension Int: ArgumentKind {
+    public init?(arg: String) {
+        self.init(arg)
+    }
+
+    public init(popArg: () throws -> (String)) throws {
+        let arg = try popArg()
+        // Not every string can be converted into an integer.
+        guard let intValue = Int(arg) else {
+            throw ArgumentParserError.typeMismatch("\(arg) is not convertible to Int")
+        }
+        self = intValue
+    }
+}
+
+extension Bool: ArgumentKind {
+    public init?(arg: String) {
+        self = true
+    }
+
+    public init(popArg: () throws -> (String)) throws {
+        // We don't need to pop here because presence of the option
+        // is enough to indicate that the bool value is true.
+        self = true
+    }
+}
+
+
+/// An abstract class representing positional or options argument.
+// FIXME: Consider renaming to Argument once we get rid of OptionsParser.
+public class CommandLineArgument: CustomStringConvertible, Hashable {
+
+    /// The name of the argument.
+    fileprivate let name: String
+
+    /// The kind of the argument
+    fileprivate let kind: ArgumentKind.Type
+
+    /// This argument can occur mutiple times.
+    fileprivate let isArray: Bool
+
+    /// The usage text associated with this argument.
+    fileprivate let usage: String?
+
+    fileprivate init(name: String, kind: ArgumentKind.Type, usage: String?, isArray: Bool) {
+        self.name = name
+        self.kind = kind
+        self.usage = usage
+        self.isArray = isArray
+    }
+
+    // MARK:- Protocol conformances
+
+    public var description: String {
+        return name + "(\(kind))"
+    }
+
+    public var hashValue: Int {
+        return name.hashValue
+    }
+
+    public static func ==(_ lhs: CommandLineArgument, _ rhs: CommandLineArgument) -> Bool {
+        return lhs.name == rhs.name &&
+               lhs.kind == rhs.kind &&
+               lhs.usage == rhs.usage &&
+               lhs.isArray == rhs.isArray
+    }
+
+    fileprivate static func isPositional(arg: String) -> Bool {
+        return !arg.hasPrefix("-")
+    }
+}
+
+/// A class representing option arguments. These are optional arguments
+/// which may or may not be provided in the command line. They are always
+/// prefixed by their name. For e.g. --verbosity true.
+public final class OptionArgument<T>: CommandLineArgument {
```

Comment:
This is kind of "kind" Normally it is kind but for arrays T is `[Kind]`
I think this can be solved now with the new extensions feature in compiler